virtualizarr/writers/icechunk.py

```diff
@@ -0,0 +1,179 @@
+        ds.variables,
+        store=store,
+        group=root_group,
+    )
+
+
+def write_variables_to_icechunk_group(
+    variables,
+    store,
+    group,
+):
+    for name, var in variables.items():
+        write_variable_to_icechunk(
+            store=store,
+            group=group,
+            name=name,
+            var=var,
+        )
+
+
+def write_variable_to_icechunk(
+    store: "IcechunkStore",
+    group: Group,
+    name: str,
+    var: Variable,
+) -> None:
+    """Write a single (possibly virtual) variable into an icechunk store"""
+
+    if isinstance(var.data, ManifestArray):
+        write_virtual_variable_to_icechunk(
+            store=store,
+            group=group,
+            name=name,
+            var=var,
+        )
+    else:
+        # TODO is writing loadable_variables just normal xarray ds.to_zarr?
+        # raise NotImplementedError()
+        print("skipping non-virtual variable", name)
+
+
+def write_virtual_variable_to_icechunk(
+    store: "IcechunkStore",
+    group: Group,
+    name: str,
+    var: Variable,
+) -> None:
+    """Write a single virtual variable into an icechunk store"""
+    ma = cast(ManifestArray, var.data)
+    zarray = ma.zarray
+
+    # creates array if it doesn't already exist
+    arr = group.require_array(
+        name=name,
+        shape=zarray.shape,
+        chunk_shape=zarray.chunks,
+        dtype=encode_dtype(zarray.dtype),
+        codecs=zarray._v3_codec_pipeline(),
+        dimension_names=var.dims,
+        fill_value=zarray.fill_value,
+        # TODO fill_value?
+    )
+
+    # TODO it would be nice if we could assign directly to the .attrs property
+    for k, v in var.attrs.items():
+        arr.attrs[k] = encode_zarr_attr_value(v)
+    arr.attrs["_ARRAY_DIMENSIONS"] = encode_zarr_attr_value(var.dims)
+
+    _encoding_keys = {"_FillValue", "missing_value", "scale_factor", "add_offset"}
+    for k, v in var.encoding.items():
+        if k in _encoding_keys:
+            arr.attrs[k] = encode_zarr_attr_value(v)
+
+    write_manifest_virtual_refs(
+        store=store,
+        group=group,
+        arr_name=name,
+        manifest=ma.manifest,
+    )
+
+
+def write_manifest_virtual_refs(
+    store: "IcechunkStore",
+    group: Group,
+    arr_name: str,
+    manifest: ChunkManifest,
+) -> None:
+    """Write all the virtual references for one array manifest at once."""
+
+    key_prefix = f"{group.name}{arr_name}"
+
+    # loop over every reference in the ChunkManifest for that array
+    # TODO inefficient: this should be replaced with something that sets all (new) references for the array at once
+    # but Icechunk need to expose a suitable API first
+    it = np.nditer(
+        [manifest._paths, manifest._offsets, manifest._lengths],  # type: ignore[arg-type]
+        flags=[
+            "refs_ok",
+            "multi_index",
+            "c_index",  # TODO is "c_index" correct? what's the convention for zarr chunk keys?
```

Comment:
If we add any test that this generates the correct result for any variable with more than one chunked dimension then we can remove this comment.